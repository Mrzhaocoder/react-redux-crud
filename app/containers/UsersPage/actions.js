/*
 *
 * UsersPage actions
 *
 * These are used both when dispatching an action into Redux, and for subscribing
 * to Redux state changes so that the Components can update accordingly.
 *
 */

import {
  LOAD_USERS,
  LOAD_USERS_ERROR,
  LOAD_USERS_SUCCESS,
} from './constants';

/**
 * Load the users, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_USERS
 */
export function loadUsers() {
  return {
    type: LOAD_USERS,
  };
}

/**
 * Dispatched when the users are loaded by the request saga
 *
 * @param  {array} users The users data
 *
 * @return {object}      An action object with a type of LOAD_USERS_SUCCESS passing the users
 */
export function usersLoaded(users) {
  return {
    type: LOAD_USERS_SUCCESS,
    users,
  };
}

/**
 * Dispatched when loading the users fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_USERS_ERROR passing the error
 */
export function usersLoadingError(error) {
  return {
    type: LOAD_USERS_ERROR,
    error,
  };
}
