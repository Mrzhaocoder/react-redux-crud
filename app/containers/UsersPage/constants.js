/*
 *
 * UsersPage constants
 *
 */

export const LOAD_USERS = 'react-redux-crud/UsersPage/LOAD_USERS';
export const LOAD_USERS_SUCCESS = 'react-redux-crud/UsersPage/LOAD_USERS_SUCCESS';
export const LOAD_USERS_ERROR = 'react-redux-crud/UsersPage/LOAD_USERS_ERROR';
