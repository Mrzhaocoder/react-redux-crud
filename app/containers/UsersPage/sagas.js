/*
 *
 * UsersPage sagas
 *
 * These watch for specific events to come from Redux (takeLatest), fetch the
 * data necessary for the given event, and then finally return it via an action
 * where any number of Components could be listening for it.
 *
 */

import { take, call, put, cancel, takeLatest } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import request from 'utils/request';

import { LOAD_USERS } from './constants';
import { usersLoaded, usersLoadingError } from './actions';

/**
 * Gets the users
 */
export function* getUsers() {
  const requestURL = 'http://localhost:3001/users';

  try {
    const users = yield call(request, requestURL);
    yield put(usersLoaded(users));
  } catch (err) {
    yield put(usersLoadingError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* usersData() {
  // Watches for LOAD_USERS actions and calls getUsers when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  const watcher = yield takeLatest(LOAD_USERS, getUsers);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [
  usersData,
];
