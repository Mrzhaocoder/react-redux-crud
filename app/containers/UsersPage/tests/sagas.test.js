import { take, put, cancel, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/utils';
import { LOCATION_CHANGE } from 'react-router-redux';

import { LOAD_USERS } from '../constants';
import { usersLoaded, usersLoadingError } from '../actions';
import { getUsers, usersData } from '../sagas';


/* eslint-disable redux-saga/yield-effects */
describe('getUsers Saga', () => {
  let getUsersGenerator;

  beforeEach(() => {
    getUsersGenerator = getUsers();

    const callDescriptor = getUsersGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the usersLoaded action if it requests the data successfully', () => {
    const users = [
      { id: 1, username: 'empurium', email: 'michael@gmail.com', address: '1 Infinite Loop', role: 'admin' },
      { id: 2, username: 'robbie', email: 'robbie@gmail.com', address: '1234 Breakpoint Street', role: 'admin' },
    ];
    const putDescriptor = getUsersGenerator.next(users).value;
    expect(putDescriptor).toEqual(put(usersLoaded(users)));
  });

  it('should call the usersLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getUsersGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(usersLoadingError(response)));
  });
});

describe('usersData Saga', () => {
  let generator;
  let taskMock;

  beforeEach(() => {
    generator = usersData();
    taskMock = createMockTask();
  });

  it('should start task to watch for LOAD_USERS action', () => {
    const takeLatestDescriptor = generator.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_USERS, getUsers));
  });

  it('should yield until LOCATION_CHANGE action', () => {
    generator.next();
    const takeDescriptor = generator.next();
    expect(takeDescriptor.value).toEqual(take(LOCATION_CHANGE));
  });

  it('should cancel after LOCATION_CHANGE action', () => {
    generator.next();
    generator.next(taskMock);
    const cancelDescriptor = generator.next();
    expect(cancelDescriptor.value).toEqual(cancel(taskMock));
  });
});
