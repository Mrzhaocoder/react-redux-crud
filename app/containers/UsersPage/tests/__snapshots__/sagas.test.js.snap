exports[`getUsers Saga should call the usersLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3001/users",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getUsers Saga should dispatch the usersLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3001/users",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
